### PR TITLE
Download CLI config to remove dashboard dependency for pctl usage

### PR DESCRIPTION
**Pull request by hiteshwani29**
### What does this PR change?

- This PR will add support in cli to download CLI configs, which will remove the dashboard dependency. This will add enhancement for the issue #38   

### Does the PR depend on any other PRs or Issues? If yes, please list them.

- NO

### Checklist

I confirm, that I have...

- [x] Read and followed the contributing guide in `CONTRIBUTING.md`
- [ ] Added tests for this PR
- [x] Formatted the code using `go fmt` (if applicable)
- [ ] Updated [documentation on the Paralus docs site](https://github.com/paralus/website/blob/main/docs) (if applicable)
- [x] Updated `CHANGELOG.md`

**Tested on local system** 
Screenshot from 2023-03-24 13-44-50

### CODE REVIEW

**Comment by hiteshwani29** [comment]
@niravparikh05 @akshay196
Please review this PR

**Review by akshay196 — CHANGES_REQUESTED** [review]
@hiteshwani29 Overall code is looking good. One major pending item:
Add ability to take user credentials as command inputs for example `--email <email-id> --password <password>`, therefore `pctl config download` can be used in scripts or CI pipelines.

I will add more comments after another code-review iteration.

**Comment by hiteshwani29** [comment]
> @hiteshwani29 Overall code is looking good. One major pending item: Add ability to take user credentials as command inputs for example `--email <email-id> --password <password>`, therefore `pctl config download` can be used in scripts or CI pipelines.
> 
> I will add more comments after another code-review iteration.

Done @akshay196 added flags email and password to take user creds as command input. 
Please review it and add your comments.

File: pkg/kratos/kratos_login.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	ctx := context.Background()
+	client := getNewAPIClient(clientURL)
+
+	// Initialize a registration flow
+	log.GetLogger().Debug("Initializing the login flow.")
+	flow, _, err := client.V0alpha2Api.InitializeSelfServiceLoginFlowWithoutBrowser(ctx).Execute()
+	if err != nil {
+		return nil, err
+	}
+	log.GetLogger().Debug("Flow id fetched successfully issued_at: %v, expires_at: %v", flow.IssuedAt, flow.ExpiresAt)
```
Comment (akshay196) [inline pkg/kratos/kratos_login.go:29]:
```suggestion
	log.GetLogger().Debugf("Flow id fetched successfully issued_at: %v, expires_at: %v", flow.IssuedAt, flow.ExpiresAt)
```

File: pkg/kratos/kratos_login.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			Identifier:         email,
+			PasswordIdentifier: &email,
+		}),
+	).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	log.GetLogger().Debug("User credentials validated successfully.")
+	log.GetLogger().Debug("User logged in successfully, session_id: %v, issued_at: %v, expires_at: %v, identity_id: %v, state: %v", result.Session.Id, result.Session.IssuedAt, result.Session.Identity.Id, result.Session.Identity.State)
```
Comment (akshay196) [inline pkg/kratos/kratos_login.go:44]:
```suggestion
	log.GetLogger().Debugf("User logged in successfully, session_id: %v, issued_at: %v, expires_at: %v, identity_id: %v, state: %v", result.Session.Id, result.Session.IssuedAt, result.Session.ExpiresAt, result.Session.Identity.Id, result.Session.Identity.State)
```

File: pkg/commands/config_download.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	_, err := url.ParseRequestURI(args[0])
+	if err != nil {
+		return err
+	}
+
+	if o.email == "" && o.password != "" || o.email != "" && o.password == "" {
+		return errors.New("please provide both email and password to login")
+	}
+
```
Comment (akshay196) [inline pkg/commands/config_download.go:49]: Another check could be checking if provided to-file path is not a directory.
```go
	if fi, err := os.Stat(o.downloadConfigFilePath); err == nil && fi.IsDir() {
		return fmt.Errorf("%s is a directory", o.downloadConfigFilePath)
	}
```

**Review by akshay196 — CHANGES_REQUESTED** [review]
This is working fine on my machine. Few minor code changes:

File: pkg/commands/config_download.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		bytePassword, err := term.ReadPassword(syscall.Stdin)
+		if err != nil {
+			return err
+		}
+		o.password = string(bytePassword)
+	}
+
+	kc, err := kratos.Login(args[0], o.email, o.password)
+	if err != nil {
+		return fmt.Errorf("failed to login. You may have entered an invalid username or password or paralus host endpoint :: %v", err.Error())
```
[the remark this answers is not in the document]
Comment (hiteshwani29) [reply pkg/commands/config_download.go]: Removed from the output and added it to the log.

File: pkg/commands/config_download.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	o.logger.Debug("Fetching CLI config for user.")
+	res, err := kc.HttpGet(fmt.Sprintf("%s/auth/v3/cli/config", args[0]))
+	if err != nil {
+		return fmt.Errorf("failed to download cli config : %v", err.Error())
+	}
+
```
[the remark this answers is not in the document]
Comment (hiteshwani29) [reply pkg/commands/config_download.go:89]: Done

File: pkg/commands/config_download.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	kc, err := kratos.Login(args[0], o.email, o.password)
+	if err != nil {
+		return fmt.Errorf("failed to login. You may have entered an invalid username or password or paralus host endpoint :: %v", err.Error())
+	}
+
+	o.logger.Debug("Fetching CLI config for user.")
+	res, err := kc.HttpGet(fmt.Sprintf("%s/auth/v3/cli/config", args[0]))
+	if err != nil {
+		return fmt.Errorf("failed to download cli config : %v", err.Error())
```
[the remark this answers is not in the document]
Comment (hiteshwani29) [reply pkg/commands/config_download.go]: Added http error in log.

File: pkg/commands/config_download.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	if o.downloadConfigFilePath == "" {
+		userHomeDir, err := os.UserHomeDir()
+		if err != nil {
+			return err
+		}
+		err = os.MkdirAll(fmt.Sprintf("%s/.paralus/cli/", userHomeDir), os.ModePerm)
+		if err != nil {
+			return err
+		}
+		o.downloadConfigFilePath = fmt.Sprintf("%s/.paralus/cli/config.json", userHomeDir)
```
[the remark this answers is not in the document]
Comment (hiteshwani29) [reply pkg/commands/config_download.go]: Done

File: pkg/commands/config_download.go (hunk 3 above)
Comment (hiteshwani29) [reply pkg/commands/config_download.go:58]: Added validation

**Comment by hiteshwani29** [comment]
@akshay196 Added requested changes.

**Comment by akshay196** [comment]
CLI config download is not working with latest Paralus release. I am getting `Data failed to match schemas in oneOf(UiNodeAttributes)` error from `InitializeSelfServiceLoginFlowWithoutBrowser().Execute()`.

I guess this is due to mismatch in kratos sdk and kratos server versions. We have updates kratos version v0.11.0 in latest Paralus release, however pctl is using krato sdk v0.10.1
Let me try updating kratos-client-go and see if it resolves.

**Comment by akshay196** [comment]
Manual tested config download command. Working fine. 💯 
[config-download-testing.md](https://github.com/paralus/cli/files/11131981/config-download-testing.md)

@niravparikh05 Feel free to comment if you've any feedback or suggestions.

**Review by akshay196 — APPROVED** [review]
LGTM.  Thanks @hiteshwani29 🎉

**Pull request merged**